File: day13_part1_2023.cpp

#include <iostream>
#include <fstream>
#include <vector>
#include <string>

struct Mirror {
    std::vector<int> Rows;
    std::vector<int> Cols;
};

Mirror parseMirror(std::vector<std::string> mirrorStr) {
    Mirror mirror;
    mirror.Rows.resize(mirrorStr.size());
    mirror.Cols.resize(mirrorStr[0].size());

    for (int y = 0; y < mirrorStr.size(); y++) {
        for (int x = 0; x < mirrorStr[0].size(); x++) {
            mirror.Rows[y] <<= 1;
            mirror.Cols[x] <<= 1;
            if (mirrorStr[y][x] == '#') {
                mirror.Rows[y]++;
                mirror.Cols[x]++;
            }
        }
    }

    return mirror;
}

int getMirrorAxis(std::vector<int> lines) {
    for (int i = 1; i < lines.size(); i++) {
        bool isMirror = true;

        for (int j = 0; isMirror && j < std::min(i, static_cast<int>(lines.size()) - i); j++) {
            if (lines[i - 1 - j] != lines[i + j]) {
                isMirror = false;
            }
        }

        if (isMirror) {
            return i;
        }
    }

    return 0;
}

int solve(std::vector<std::string> input) {
    std::vector<Mirror> mirrors;

    std::vector<std::string> mirrorStr;
    for (auto line : input) {
        if (line == "") {
            mirrors.push_back(parseMirror(mirrorStr));
            mirrorStr.clear();
        } else {
            mirrorStr.push_back(line);
        }
    }
    mirrors.push_back(parseMirror(mirrorStr));

    int res = 0;
    for (auto mirror : mirrors) {
        res += getMirrorAxis(mirror.Cols);
        res += getMirrorAxis(mirror.Rows) * 100;
    }
    return res;
}

int main() {
    std::ifstream inputFile("input.txt");
    std::vector<std::string> input;
    std::string line;

    while (std::getline(inputFile, line)) {
        input.push_back(line);
    }

    std::cout << solve(input) << std::endl;

    return 0;
}

File: haskell/day13_part1_2023.hs

import System.IO
import Data.Bits
import Data.List
import Data.Maybe (fromMaybe)

parseMirror :: [String] -> ([Int], [Int])
parseMirror mirrorStr = (rows, cols)
  where
    rows = map rowToInt mirrorStr
    cols = map colToInt [0 .. (length (head mirrorStr) - 1)]
    rowToInt line = foldl (\acc c -> (acc `shiftL` 1) + (if c == '#' then 1 else 0)) 0 line
    colToInt x = foldl (\acc line -> (acc `shiftL` 1) + (if line !! x == '#' then 1 else 0)) 0 mirrorStr

getMirrorAxis :: [Int] -> Int
getMirrorAxis lines = fromMaybe 0 $ find (\i -> all (\j -> lines !! (i - 1 - j) == lines !! (i + j)) [0 .. min (i - 1) (length lines - i - 1)]) [1 .. (length lines - 1)]

getMirrorAxisWithOneSmudge :: [Int] -> Int
getMirrorAxisWithOneSmudge lines = fromMaybe 0 $ find (\i -> (== 1) $ sum $ map (\j -> bitDiff (lines !! (i - 1 - j)) (lines !! (i + j))) [0 .. min (i - 1) (length lines - i - 1)]) [1 .. (length lines - 1)]
  where
    bitDiff x y = if (xor x y) .&. ((xor x y) - 1) == 0 then if x == y then 0 else 1 else 0

solve :: [String] -> Int
solve input = sum $ map processMirror $ splitMirrors input
  where
    splitMirrors :: [String] -> [[String]]
    splitMirrors [] = []
    splitMirrors input = mirror : splitMirrors rest
      where
        (mirrorLines, restWithNewline) = break (== "") input
        mirror = mirrorLines
        rest = drop 1 restWithNewline

    processMirror :: [String] -> Int
    processMirror mirrorStr = cols + rows * 100
      where
        (rowsLines, colsLines) = parseMirror mirrorStr
        cols = getMirrorAxis colsLines
        rows = getMirrorAxis rowsLines

main :: IO ()
main = do
  contents <- readFile "input.txt"
  let inputLines = lines contents
  print $ solve inputLines
